### PR TITLE
fix tokamax splash

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

Protect tokamax's splash attention call under a flag `use_tokamax_splash`

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/455572237

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

unit tests

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: generated_requirements/cuda12-requirements.txt [hunk 1]
```diff
@@ -1,248 +1,261 @@
...
-wrapt>=1.17.3
-xxhash>=3.6.0
-yarl>=1.22.0
-zipp>=3.23.0
-zstandard>=0.23.0
+absl-py==2.3.1
+aiofiles==25.1.0
+aiohappyeyeballs==2.6.1
+aiohttp==3.13.1
+aiosignal==1.4.0
```
Comment (bvandermoon) [inline generated_requirements/cuda12-requirements.txt]: Why update the GPU dependencies?

Comment (khatwanimohit) [reply generated_requirements/cuda12-requirements.txt]: tokamax was not added as gpu-dependency in #2531. now that JAXAII image and MaxText package have upgraded to 0.8.0 we can remove conditional imports of tokamax and add it as a deps everywhere

Comment (bvandermoon) [reply generated_requirements/cuda12-requirements.txt]: Why does this use all `==` instead of `>=` now though?

File: src/MaxText/layers/attention_op.py [hunk 2]
```diff
@@ -31,12 +31,10 @@
...
-if jax.__version__ < "0.8.0":
-  from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_kernel
-  from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_mask
-else:
-  from tokamax._src.ops.experimental.tpu.splash_attention import splash_attention_kernel
-  from tokamax._src.ops.experimental.tpu.splash_attention import splash_attention_mask
+from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_kernel
+from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_mask
+
+from tokamax._src.ops.experimental.tpu.splash_attention import splash_attention_kernel as tokamax_splash_kernel
```
Comment (suexu1025) [inline src/MaxText/layers/attention_op.py:37]: we also need to import mask from tokamax

Comment (khatwanimohit) [reply src/MaxText/layers/attention_op.py:37]: done

File: generated_requirements/cuda12-requirements.txt (hunk 1 above)
Comment (khatwanimohit) [reply generated_requirements/cuda12-requirements.txt]: discussed offline

File: generated_requirements/cuda12-requirements.txt [hunk 3]
```diff
@@ -1,248 +1,261 @@
...
+wadler-lindig==0.1.7
+websockets==15.0.1
+werkzeug==3.1.3
+wheel==0.45.1
+wrapt==2.0.0
+xprof==2.20.7
+xxhash==3.6.0
+yarl==1.22.0
+zipp==3.23.0
+zstandard==0.25.0
```
Comment (bvandermoon) [inline generated_requirements/cuda12-requirements.txt]: Are there corresponding `base_requirements` changes for this file?

Comment (khatwanimohit) [reply generated_requirements/cuda12-requirements.txt]: yes, these requirements are generated using command: `seed-env --local-requirements=base_requirements/gpu-base-requirements.txt --host-name=MaxText --seed-commit=459cb056418de7a56c9da0a2842406a58b75e4a3 --python-version=3.12  --output-dir=generated_cuda12_artifacts --hardware=cuda12`

Comment (bvandermoon) [reply generated_requirements/cuda12-requirements.txt]: Thanks Mohit. Can you please also add the base requirements changes in this PR?

Comment (khatwanimohit) [reply generated_requirements/cuda12-requirements.txt]: there were no changes done to the base requirements file. I guess these deps are the second order dependency added by seed-env

Comment (bvandermoon) [reply generated_requirements/cuda12-requirements.txt]: @kanglant can you weigh in on what's happening here/why this is using `==` now instead of `>=`?

Comment (kanglant) [reply generated_requirements/cuda12-requirements.txt]: @khatwanimohit could you rerun the `seed-env` command with the `--requirements-txt=cuda12-requirements.txt` flag? It will generate a requirements lock file containing lower bounds. We prefer to use lower bounds rather than fixed versions for greater flexibility if needed.

Comment (khatwanimohit) [reply generated_requirements/cuda12-requirements.txt]: thanks @kanglant , I re-ran with that flag and now see the deps are with `>=`

File: src/MaxText/layers/attention_op.py
Comment (suexu1025) [inline src/MaxText/layers/attention_op.py:1142]: @khatwanimohit  the mask here also needs to change to tokamax version.

Comment (suexu1025) [inline src/MaxText/layers/attention_op.py:1144]: same for splash_attention_mask

Comment (suexu1025) [inline src/MaxText/layers/attention_op.py:1155]: same for splash_attention_mask

File: tests/integration_tests/train_tests.py [hunk 4]
```diff
@@ -142,6 +142,11 @@ class TrainTests(unittest.TestCase):
   def test_tpu_base(self):
     train_main(TrainTests.CONFIGS["base"])
 
+  @pytest.mark.integration_test
+  @pytest.mark.tpu_only
+  def test_tpu_tokamax(self):
+    train_main(TrainTests.CONFIGS["base"] + ["use_tokamax_splash=true"])
```
Comment (suexu1025) [inline tests/integration_tests/train_tests.py:148]: can we specify attention=flash here?

File: src/MaxText/layers/attention_op.py
Comment (khatwanimohit) [reply src/MaxText/layers/attention_op.py:1142]: tokamax's splash attention mask is basically a fork of the jax's splash attention mask, There is no diff in between them except MultiHeadMask. See diff here: https://diff.googleplex.com/#key=fusskAKvgNJw

File: src/MaxText/layers/attention_op.py [hunk 5]
```diff
@@ -1291,7 +1288,7 @@ def wrap_flash_attention(
       if version.parse(jax.__version__) < version.parse("0.7.2.dev20250824"):
```
Comment (RissyRan) [inline src/MaxText/layers/attention_op.py:1288]: @Rohan-Bierneni we are running 0.8.0 right? I think this `if` could be removed.

**Review by RissyRan — APPROVED** [review]
Thanks Mohit for the fix! Just one minor comment, and I could sent out PR separately as well.
